More detailed prototyping of Dataset.

SimonHeybrock opened the pull request:
Work-in-progress finding and implementing concrete solutions outlined and suggested in the previous iteration.

See mainly `doc/type-erased-prototype.md` for a description. A list of notable changes and to-do list with items to be covered in this iteration can be found towards the end of the document.

SimonHeybrock committed:
First part of cleanup, based on previously updated notes.

SimonHeybrock committed:
Start adding DataArray, to replace and extend ColumnHandle.

SimonHeybrock committed:
Basic DataArray tests.

SimonHeybrock committed:
Easier DataArray init.

SimonHeybrock committed:
Add Dimensions class.

SimonHeybrock committed:
Support only a single ragged dimension.

SimonHeybrock committed:
Ragged support in DataArray via Dimensions.

SimonHeybrock committed:
Add note regarding resizing DataArray.

SimonHeybrock committed:
Add concatenate for DataArray.

SimonHeybrock committed:
Unanswered questions regarding variables depending on others.

SimonHeybrock committed:
Small update of prototype notes.

SimonHeybrock committed:
Changes in chaotic state.

SimonHeybrock committed:
Cleanup in DataArray.

SimonHeybrock committed:
Start using DataArray in Dataset.

SimonHeybrock committed:
DatasetIterator works with latest Dataset changes.

SimonHeybrock committed:
Use updated Dataset API in Histogram test.

SimonHeybrock committed:
Cleanup in Dataset.

SimonHeybrock committed:
First sketch of concatenate for Dataset.

SimonHeybrock committed:
Basic support for units in DataArray.

SimonHeybrock committed:
Rename Dataset::add to Dataset::insert.

SimonHeybrock committed:
Add Dataset::insertAsEdge.

SimonHeybrock committed:
More tests for edge support in Dataset.

SimonHeybrock committed:
Problems with DatasetIterator.

SimonHeybrock committed:
DatasetIterator: Do not iterate Dimensions not spanned by variables.

SimonHeybrock committed:
Refactor to improve DatasetIterator support for edge variables.

SimonHeybrock committed:
Fix more issues with bin edge iteration.

SimonHeybrock committed:
Named getters for DatasetIterator.

SimonHeybrock committed:
Proper exception message.

SimonHeybrock committed:
Access to bins via DatasetIterator.

SimonHeybrock committed:
Cleanup.

SimonHeybrock committed:
Remove test code that is now implemented.

SimonHeybrock committed:
Rename DatasetIterator to DatasetView.

SimonHeybrock committed:
Update notes.

SimonHeybrock committed:
First hacky support for Histogram in DatasetView.

SimonHeybrock committed:
Small progress for Histogram access.

SimonHeybrock committed:
Confusion shall be my epitaph.

SimonHeybrock committed:
Expand notes on unsolved problems.

SimonHeybrock committed:
Fix build of benchmarks after API changes.

SimonHeybrock committed:
Improving tag ID handling and preparations for coord/data var split.

SimonHeybrock committed:
Rename Variable tag to Data.

SimonHeybrock committed:
Validation in Dataset::insert.

SimonHeybrock committed:
Dataset access with name.

SimonHeybrock committed:
Test DatasetView fail on duplicate tags.

SimonHeybrock committed:
Cleanup in DatasetView.

SimonHeybrock committed:
More cleanup in DatasetView.

SimonHeybrock committed:
Basic support for accessing named variables via DatasetIterator.

SimonHeybrock committed:
Cleanup and update dataset prototype notes.

SimonHeybrock committed:
Extend notes with TODO list.

SimonHeybrock committed:
Remove outdated commented code.

SimonHeybrock committed:
Benchmarks related to common Mantid memory access patterns.

SimonHeybrock committed:
Benchmarks related to cow_ptr::access().

SimonHeybrock committed:
Update notes.

SimonHeybrock committed:
Add more benchmarks related to multi-dimensional indexing.

SimonHeybrock committed:
Small refactor in LinearSubindex.

SimonHeybrock committed:
Add random-access iterator for DatasetView.

SimonHeybrock committed:
Getter mixins for DatasetView iterator.

SimonHeybrock committed:
Small improvements(?) for iterator code.

OwenArnold reviewed (commented):
Aside from the specific comments make, I think it would be worth investing in cmake + ci to give greater confidence that these features are working outside of your local environment. I was unable to build the prototype with apple clang for GCC, and it looks like there are currently some genuine errors in the source. This will be especially important with more people looking at the codebase.

I haven't reviewed the view parts of the code yet, and would like to do that.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -63,8 +63,8 @@ We have several options:
    Furthermore it provides an elegant way of dealing with variables that do not share all their dimensions.
-   - The view/iterator could iterate either all (applicable) dimensions, a specific dimensions, or all but certain dimensions, depending on the needs to the specific client code.
-   - See [`test/dataset_iterator_test.cpp`](../src/type-erased-prototype/test/dataset_iterator_test.cpp) for examples.
+   - The view/iterator could iterate either all (applicable) dimensions, a specific dimension, or all but certain dimensions, depending on the needs to the specific client code.
```
The iterator would be very generic and would know about histogram. If we iterate histograms, we would return a reference to a `Histogram`, which will be different from the current `HistogramData::Histogram`. Unless it is copied it would reference data in the `Dataset`. By current idea regarding handling frequencies/counts/std-dev/variance is via the *unit* of Y and E.

SimonHeybrock commented on src/type-erased-prototype/variable.h:
```diff
@@ -0,0 +1,173 @@
+  struct Histogram {};
+
+  using tags = std::tuple<Tof, Value, Error, Int, DimensionSize, Histogram>;
+};
```
In most cases it could use one of the generic tags like `Int` or `Value` (double) and get identified by its name. If it is different data there are two options:
- Add a new tag in case it is something that will be used frequently.
- Have dummy tags like `Custom` (or `Custom1`, `Custom2`?).

SimonHeybrock commented on src/type-erased-prototype/data_array.h:
```diff
@@ -0,0 +1,184 @@
+#include "index.h"
+#include "variable.h"
+
+class DataArrayConcept {
```
It is used a the type-erased handle for all variable types. `DataArrayModel` is a template and thus depends on the actual held type.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,86 @@ if (d.contains<BinMask>())
+
+Outstanding tasks:
+
+- Full sample implementation of `concatenate` and `operator+` for `Dataset`, demonstrating the full suite of required features from `Dataset` and how they could be implemented.
```
I think the behaviour of such operations with `Dataset` was very unclear in v1. With the changes since then I think I developed a better understanding --- leading to the distinction between coordinate variables and data variables (based on whether operations match them for compatibility checks or transform them).

SimonHeybrock commented on src/type-erased-prototype/dataset.h:
```diff
@@ -1,145 +1,183 @@
-  const std::map<Dimension, gsl::index> &dimensions() const {
-    return m_dimensions;
-  }
+  Dimensions m_dimensions;
```
There are a couple of advantages for storing the dimension extents also in `DataArray` (different length for bin edges, stand-alone `DataArray` used in initialization, easier implementation of operations due to better encapsulation).

`Dataset` synchronizes the dimensions as part of validation: The dimensions of the held `DataArray` instances must match, with some flexibility (if length is off by one it is bin edges, dimension order can differ).

SimonHeybrock commented on src/type-erased-prototype/dimensions.cpp:
```diff
@@ -0,0 +1,232 @@
+}
+Dimensions::Dimensions(const Dimensions &other) : m_dims(other.m_dims) {
+  if (other.m_raggedDim)
+    m_raggedDim = std::make_unique<DataArray>(*other.m_raggedDim);
```
Ok. Note that since I added basic ragged-dimension support I have struggled a bit with making it work seamlessly. The main problem appears to be our multi-dimensional iteration via `DatasetView`. Random access iterators do not seem to be possible if we want to support ragged dimensions. Without random access multi-threading and also other code becomes slightly more cumbersome to use.

SimonHeybrock commented on src/type-erased-prototype/dimensions.cpp:
```diff
@@ -0,0 +1,232 @@
+Dimensions concatenate(const Dimension dim, const Dimensions &dims1,
+                       const Dimensions &dims2) {
+  if (dims1.contains(dim) && dims2.contains(dim)) {
+    // - all dimension labels must match and have same order.
```
https://github.com/mantidproject/workspace-sandbox/pull/4/files/1253cda5281b4c5faac377ac18bb0cd7ae213651#diff-0e63ae6420c4d4f8342b5af8c0c3ce79R22.

SimonHeybrock commented:
@OwenArnold The unsolved problems I currently see with the iterator are:

1. In the non-ragged-dimension case: The implemented random-access iterator is relatively expensive. In benchmarks like https://github.com/mantidproject/workspace-sandbox/pull/4/files#diff-33229b4f09798097a501b57fb858c763R102 which require, e.g., 3 doubles = 24 Byte per iterator position, we are limited by the index math to a sustained memory bandwidth of 2 to 4 GB/s (on a system that can support 20 to 30 GB/s in other benchmarks). Note that this is even *with* multi threading. Some improvements to the iterator are still possible but it is not clear whether it can ever the considered cheap with so little data. That is, it is not clear whether its performance is sufficient when implementing operations for large datasets on the lowest level. Iterating on larger constructs like `Histogram` is not affected.

2. In the ragged-dimension case: Providing a random-access iterator is not feasible. We either have access that is not constant-time (involving search through ragged length information) or expensive constant-time access by building a large index map, e.g., building/allocating an `unordered_map`. Providing a *forward* iterator should be possible and more efficient. How does this influence usability?

3. Iterating multiple datasets at the same time: Consider implementing operations involving two `Datasets`. We need to iterate both datasets in precisely the same way, i.e, in the same order of dimensions, and in all dimensions, even if one of the datasets does not depend on one of those dimensions. One option is to copy the variables in questions into a joint `Dataset` first such that a single iterator is sufficient. Otherwise we need to come up with a solution to creating synchronized iterators.

OwenArnold commented on src/type-erased-prototype/benchmark/dataset_view_benchmark.cpp:
```diff
@@ -0,0 +1,156 @@
+#pragma omp parallel for num_threads(state.range(1)) schedule (static,10000)
+    for (gsl::index i = 0; i < count; ++i) {
+      auto item = view.begin() + i;
+      item->get<Data::Value>() += item->get<const Data::Error>();
```
This looks to be an interesting benchmark. When I can build/run the code locally I'd like to run this myself. Presumably all time spent in `LinearSubIndex`

SimonHeybrock replied:
I think the biggest impact is the long latency from the `/` and `%` operations to compute the indices for each dimension. `LinearSubIndex` is also relevant but not the leading contribution.

OwenArnold commented on src/type-erased-prototype/dataset_view.h:
```diff
@@ -0,0 +1,426 @@
+  private:
+    friend class iterator;
+    Item(const Item &other) = default;
+    std::array<gsl::index, 3> m_indices;
```
So fixed for 3 dimension?

SimonHeybrock replied:
3 or less. This was just a quick test. I saw an improvement over using dynamically allocated memory, so I just wanted something for testing. The limit on the dimension count would of course be different in the end and the implementation may be completely different.

OwenArnold reviewed (commented):
Regarding your point (3) - You had in mind something like CompareWorkspaces? Two datasets that are the same have same DataArrays but in different order? Can we artificially order columns before iteration?

SimonHeybrock committed:
Update build instructions.

SimonHeybrock committed:
Fix gcc7 compilation errors.

SimonHeybrock committed:
A faster multi-dimensional indexing mechanism?

The simple example can do 1 Giga indices/s with gcc5 on Haswell.

SimonHeybrock committed:
End for MultiIndex. Not needed?

SimonHeybrock committed:
Mostly working and faster DatasetView iterator.

SimonHeybrock committed:
Add another todo.

SimonHeybrock committed:
Remove use of old iterator code.

SimonHeybrock committed:
WIP removing old iterator code.

SimonHeybrock committed:
Remove old multidimensional index code.

SimonHeybrock committed:
Small improvement of histogram test.

SimonHeybrock committed:
Re-enable and clean up bin code.

SimonHeybrock committed:
Small fix in MultiIndex.

SimonHeybrock committed:
Support transposed variables.

SimonHeybrock committed:
Remove wrong comment.

SimonHeybrock committed:
Clarify that Dataset::iterator items cannot be copied.

SimonHeybrock committed:
Remove old and unused histogram test code.

SimonHeybrock committed:
Fix issue with 3D or higher indices in MultiIndex.

SimonHeybrock committed:
More efficient iterator for use with OpenMP.

SimonHeybrock committed:
Working example for "virtual" access to spectrum positions.

SimonHeybrock committed:
Update todo list.

SimonHeybrock committed:
Do not require const in DatasetView::iterator::get and related cleanup.

SimonHeybrock committed:
Small cleanup in DatasetView.

SimonHeybrock committed:
Add const variants of Dataset::get.

SimonHeybrock committed:
Minor cleanup.

SimonHeybrock committed:
Add the GSL as CMake external project.

SimonHeybrock committed:
Use gsl::span to return data from DataArray and Dataset.

SimonHeybrock committed:
Cleanup unnecessary helper type structs.

SimonHeybrock committed:
Comment some tests.

SimonHeybrock committed:
Remove need to explicitly specify all element return types.

SimonHeybrock committed:
Use gsl::index from GSL.

SimonHeybrock committed:
Add tests for COW in DatasetView.

SimonHeybrock committed:
Thoughts on on-the-fly binning for events.

SimonHeybrock committed:
Some thoughts on Histogram.

SimonHeybrock committed:
Minor doc additions.

SimonHeybrock committed:
Rename header file.

SimonHeybrock committed:
Rename DataArray to Variable.

SimonHeybrock committed:
Some CMakeLists cleanup.

SimonHeybrock committed:
Add Variable::operator==.

SimonHeybrock committed:
Basic working version of concatenate for Dataset.

SimonHeybrock committed:
Add basic Variable::operator+=.

SimonHeybrock committed:
Add basic Dataset::operator+=.

SimonHeybrock committed:
Handle content mismatch in Dataset::operator+=.

SimonHeybrock committed:
Support broadcast and transpose in Variable::operator+=.

SimonHeybrock committed:
Add test for Dataset::operator+= broadcast and transpose.

SimonHeybrock committed:
Remove DatasetView Histogram mechanism as preparation for replacement.

SimonHeybrock committed:
Add support for nested DatasetView, to replace old Histogram mechanism.

SimonHeybrock committed:
Add another tests for nesting DatasetView.

SimonHeybrock committed:
Remove unused class Slab.

SimonHeybrock committed:
Working nested DatasetView, with distinct dimensionality.

SimonHeybrock committed:
Fix behavior with bad const-ness.

SimonHeybrock committed:
Add test of copy-on-write in nested case.

SimonHeybrock committed:
Cleanup some Slab/Histogram related testing disabled previously.

SimonHeybrock committed:
Rename member with outdated name.

SimonHeybrock committed:
Fix Bin for multidimensional case and support non-inner dimension.

SimonHeybrock committed:
Make bins work with nested DatasetView.

SimonHeybrock committed:
Create MultiIndex in DatasetView instead of iterator.

SimonHeybrock committed:
Remove duplicate 'lib' name from library.

SimonHeybrock committed:
Move some specializations in DatasetView out of header.

SimonHeybrock committed:
Small cleanup in DatasetView.

SimonHeybrock committed:
Forbit empty names of data variables.

SimonHeybrock committed:
Revert "Forbit empty names of data variables."

This reverts commit f7a5ac35f6be9edf0fe243ed5a2d14dab72132d7.

SimonHeybrock committed:
Move Dataset method implementations into source.

SimonHeybrock committed:
Add benchmarks for iterating nested DatasetView.

BM_DatasetView_multi_column_mixed_dimension_nested/8
44692 ns      44694 ns      12709   4.00087GB/s   174.801k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/16
77543 ns      77545 ns       9148   4.61186GB/s   201.495k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/32
149577 ns     149582 ns       4662   4.78169GB/s   208.915k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/64
294022 ns     294033 ns       2370   4.86515GB/s   212.562k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/128
578500 ns     578524 ns       1202   4.94538GB/s   216.067k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/256
1187829 ns    1187877 ns        603   4.81704GB/s    210.46k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/512
2324691 ns    2324778 ns        284   4.92266GB/s   215.074k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/1024
4803490 ns    4803651 ns        148   4.76475GB/s   208.175k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/2048
9771900 ns    9772248 ns         72   4.68432GB/s   204.661k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/4096
19297616 ns   19298163 ns         36   4.74412GB/s   207.274k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/8192
39850263 ns   39851272 ns         18   4.59472GB/s   200.746k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/16384
79696687 ns   79696875 ns          9   4.59505GB/s   200.761k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/32768
159102705 ns  159105477 ns          4   4.60337GB/s   201.124k items/s
BM_DatasetView_multi_column_mixed_dimension_nested/65536
313834001 ns  313836881 ns          2   4.66753GB/s   203.928k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/8
47991 ns      47994 ns      14630   3.72579GB/s   162.782k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/16
94582 ns      94586 ns       7376     3.781GB/s   165.194k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/32
238685 ns     238696 ns       2919   2.99651GB/s    130.92k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/64
458562 ns     458578 ns       1539   3.11945GB/s   136.291k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/128
959128 ns     959167 ns        727   2.98282GB/s   130.321k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/256
2111115 ns    2111190 ns        332   2.71034GB/s   118.417k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/512
7901348 ns    7901645 ns         88   1.44832GB/s    63.278k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/1024
18477144 ns   18477715 ns         38   1.23869GB/s   54.1192k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/2048
42858872 ns   42859861 ns         16   1093.68MB/s   46.6637k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/4096
95346096 ns   95347448 ns          8   983.246MB/s   41.9518k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/8192
225446391 ns  225454644 ns          3   831.653MB/s   35.4839k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/16384
707623766 ns  707646135 ns          1   529.926MB/s   22.6102k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/32768
1462016636 ns 1462063322 ns          1   512.974MB/s   21.8869k items/s
BM_DatasetView_multi_column_mixed_dimension_nested_transpose/65536
3117528486 ns 3117645511 ns          1   481.132MB/s   20.5283k items/s

SimonHeybrock committed:
Rename Data::Error to Data::Variance.

SimonHeybrock committed:
Add Data::StdDev as derived variable from Data::Variance.

SimonHeybrock committed:
Temporarily added CXX flags.

SimonHeybrock committed:
Slight compile time improvement.

SimonHeybrock committed:
Add default units and Variable::operator*=.

SimonHeybrock committed:
Dataset::operator*=.

SimonHeybrock committed:
Test and fix unit support in Dataset::operator*=.

SimonHeybrock committed:
Add example of Dataset replacing TableWorkspace.

SimonHeybrock committed:
Greatly improved iteration performance by avoiding allocations.

This mainly affects the nested case. I see a 2x to 3x speedup when
operating out of cache with multiple threads. On my machine only
marginal speedup when operating out of memory.

SimonHeybrock committed:
Remove redundant check in MultiIndex.

SimonHeybrock committed:
Use int32_t instead of gsl::index for small indices in MultiIndex.

Does not really seem to matter for performance.

SimonHeybrock committed:
Improve const-ness in DatasetView.

SimonHeybrock committed:
Cleanup in DatasetView that breaks performance for some reason.

It seems totally unrelated, but this change stops inlining of
iterator::increment(), causeing a 2x to 3x slowdown.

SimonHeybrock committed:
Cleanup and change in MultiIndex to help with inlining.

Moving extra code to separate function makes inlining work again,
performance is restored.

SimonHeybrock committed:
Add iterators over variables in Dataset.

SimonHeybrock committed:
Some fixes in DatasetView for access with name.

SimonHeybrock committed:
More flexible concatenate.

SimonHeybrock committed:
Work in progress example for Workspace2D generalization.

SimonHeybrock committed:
Add notes on slicing and memory pool.

SimonHeybrock committed:
Comment in slicing in Workspace2D test.

SimonHeybrock committed:
Add slice function for Variable.

SimonHeybrock committed:
Add `slice` for Dataset.

SimonHeybrock committed:
Cleanup Workspace2D example.

SimonHeybrock committed:
Use slice in TableWorkspace example.

SimonHeybrock committed:
Update notes.

SimonHeybrock committed:
Add missing file for DatasetIndex.

SimonHeybrock committed:
Add example with scanning instrument.

SimonHeybrock committed:
Details on instrument in Dataset.

SimonHeybrock committed:
Update notes.

SimonHeybrock committed:
Support DatasetView with const Dataset.

SimonHeybrock committed:
Update dimensions in Dataset::erase.

SimonHeybrock committed:
Unify arithmetic helpers in implementation of Variable operations.

SimonHeybrock committed:
Unify code in Variable arithmetic operations.

SimonHeybrock committed:
Add operator-=.

SimonHeybrock committed:
Formatting.

SimonHeybrock committed:
Add thoughts on intra-dataset operations.

SimonHeybrock committed:
Avoid copy if rvalue is passed to operator+.

SimonHeybrock committed:
Add Dataset::extract.

SimonHeybrock committed:
Add Dataset::merge.

SimonHeybrock committed:
Add example of subtracting variables within same Dataset.

SimonHeybrock committed:
Update notes.

SimonHeybrock committed:
Remove printf from test.

SimonHeybrock committed:
Cleanup.

SimonHeybrock committed:
Add masking example.

SimonHeybrock committed:
Extend masking example.

SimonHeybrock committed:
Cheaper Variable::setDimension and histogram-related benchmarks.

SimonHeybrock commented:
@OwenArnold and other reviewers:
Since the last review, some internals have been refactored significantly, so an incremental review is probably not beneficial. I recommend reviewing all changes in this PR as a union. Not all implementation details need to be reviewed as this point. In many places the implementation is preliminary and unit test coverage is in general very lacking --- the main function of the tests present at the current stage is to *demonstrate features*.

As entry points for review, I recommend looking at the following:

- As before, `doc/type-erased-prototype.md` gives an overview. Changes and open problems are summarized and discussed towards the end.
- `TableWorkspace_test.cpp`
- `Workspace2D_test.cpp`
- `dataset_test.cpp`
- `dataset_view_test.cpp`

The test cases in the above files demonstrate most key features.

Benchmarks of interest are:
- `dataset_view_benchmark.cpp` demonstrating performance related to the multi-dimensional iteration.
- `BM_Dataset_as_Histogram` is an initial attempt of investigating whether we can use `Dataset` to replace `Histogram` as well.

SimonHeybrock committed:
Speedup use of small Datasets using boost::container::small_vector.

Before:
--------------------------------------------------------------------------
Benchmark                                   Time           CPU
Iterations
--------------------------------------------------------------------------
BM_Dataset_as_Histogram/1           855499436 ns  855515634 ns
1   17.8358MB/s   1.11474M items/s
BM_Dataset_as_Histogram/2           677042194 ns  677051454 ns
1   45.0742MB/s   1.40857M items/s
BM_Dataset_as_Histogram/4           720279370 ns  720313652 ns
1   84.7341MB/s   1.32397M items/s
BM_Dataset_as_Histogram/8           688888365 ns  688911154 ns
1   177.193MB/s   1.38432M items/s
BM_Dataset_as_Histogram/16          476449959 ns  476466186 ns
2   320.249MB/s   1.25097M items/s
BM_Dataset_as_Histogram/32          242147547 ns  242156465 ns
3   630.121MB/s   1.23071M items/s
BM_Dataset_as_Histogram/64          130708321 ns  130712010 ns
5      1.14GB/s      1.14M items/s
BM_Dataset_as_Histogram/128          86234799 ns   86236983 ns
8   1.72793GB/s   884.701k items/s
BM_Dataset_as_Histogram/256          54498356 ns   54499688 ns
13   2.73414GB/s   699.939k items/s
BM_Dataset_as_Histogram/512          36884830 ns   36886568 ns
19   4.03967GB/s   517.078k items/s
BM_Dataset_as_Histogram/1024         27618638 ns   27619410 ns
25   5.39483GB/s   345.269k items/s
BM_Dataset_as_Histogram/2048         24509226 ns   24509847 ns
30   6.07865GB/s   194.517k items/s
BM_Dataset_as_Histogram/4096         20762491 ns   20763515 ns
34   7.17541GB/s   114.807k items/s
BM_Dataset_as_Histogram/8192         20873287 ns   20874348 ns
33   7.13439GB/s   57.0751k items/s
BM_Dataset_as_Histogram/16384        22494378 ns   22495013 ns
31   6.62039GB/s   26.4816k items/s
BM_Dataset_as_Histogram/32768        22409861 ns   22410438 ns
31   6.64538GB/s   13.2908k items/s
BM_Dataset_as_Histogram_with_slice  146612269 ns  146616046 ns
5   1040.73MB/s   66.6068k items/s

After:
--------------------------------------------------------------------------
Benchmark                                   Time           CPU
Iterations
--------------------------------------------------------------------------
BM_Dataset_as_Histogram/1           395761438 ns  395762207 ns
2   38.5554MB/s   2.40972M items/s
BM_Dataset_as_Histogram/2           384365167 ns  384363698 ns
2   79.3977MB/s   2.48118M items/s
BM_Dataset_as_Histogram/4           384675484 ns  384690412 ns
2    158.66MB/s   2.47907M items/s
BM_Dataset_as_Histogram/8           375414119 ns  375413317 ns
2   325.162MB/s   2.54033M items/s
BM_Dataset_as_Histogram/16          245245414 ns  245245308 ns
3   622.185MB/s   2.43041M items/s
BM_Dataset_as_Histogram/32          130681823 ns  130679916 ns
5   1.14028GB/s   2.28056M items/s
BM_Dataset_as_Histogram/64           85968954 ns   85968666 ns
8   1.73332GB/s   1.73332M items/s
BM_Dataset_as_Histogram/128          69022921 ns   69025329 ns
11    2.1588GB/s    1105.3k items/s
BM_Dataset_as_Histogram/256          48657493 ns   48659227 ns
14   3.06231GB/s   783.952k items/s
BM_Dataset_as_Histogram/512          34422514 ns   34423815 ns
20   4.32868GB/s   554.071k items/s
BM_Dataset_as_Histogram/1024         27167857 ns   27167816 ns
27   5.48451GB/s   351.008k items/s
BM_Dataset_as_Histogram/2048         22364413 ns   22365212 ns
31   6.66154GB/s   213.169k items/s
BM_Dataset_as_Histogram/4096         20348884 ns   20348570 ns
34   7.32173GB/s   117.148k items/s
BM_Dataset_as_Histogram/8192         20206221 ns   20206979 ns
35   7.37002GB/s   58.9601k items/s
BM_Dataset_as_Histogram/16384        21644786 ns   21645820 ns
32   6.88012GB/s   27.5205k items/s
BM_Dataset_as_Histogram/32768        21832418 ns   21832296 ns
32   6.82135GB/s   13.6427k items/s
BM_Dataset_as_Histogram_with_slice  130711187 ns  130717224 ns
4   1.13995GB/s    74.708k items/s

SimonHeybrock committed:
More small Dataset performance optimizations.

Avoiding even more allocations, better packing. More optimizations
should be possible, e.g., there is a lot of cost in repeatedly
cosntructing MultiIndex. Could optimize construction, or just store the
"natural" index in each Variable?

--------------------------------------------------------------------------
Benchmark                                   Time           CPU
Iterations
--------------------------------------------------------------------------
BM_Dataset_as_Histogram/1           289993818 ns  290003156 ns
3   52.6159MB/s    3.2885M items/s
BM_Dataset_as_Histogram/2           292629258 ns  292642654 ns
2   104.283MB/s   3.25884M items/s
BM_Dataset_as_Histogram/4           288159460 ns  288169323 ns
3   211.803MB/s   3.30942M items/s
BM_Dataset_as_Histogram/8           284058116 ns  284065755 ns
2   429.726MB/s   3.35723M items/s
BM_Dataset_as_Histogram/16          186663223 ns  186669963 ns
4   817.421MB/s   3.19305M items/s
BM_Dataset_as_Histogram/32          111703597 ns  111707224 ns
7   1.33395GB/s    2.6679M items/s
BM_Dataset_as_Histogram/64           75617765 ns   75619077 ns
9   1.97056GB/s   1.97056M items/s
BM_Dataset_as_Histogram/128          55178475 ns   55180211 ns
13   2.70045GB/s   1.35023M items/s
BM_Dataset_as_Histogram/256          46153422 ns   46154920 ns
15   3.22847GB/s   826.488k items/s
BM_Dataset_as_Histogram/512          31741787 ns   31742790 ns
22   4.69429GB/s   600.868k items/s
BM_Dataset_as_Histogram/1024         25246736 ns   25247923 ns
24   5.90156GB/s     377.7k items/s
BM_Dataset_as_Histogram/2048         22481351 ns   22482119 ns
32    6.6269GB/s   212.061k items/s
BM_Dataset_as_Histogram/4096         20436080 ns   20436511 ns
34   7.29023GB/s   116.644k items/s
BM_Dataset_as_Histogram/8192         20651856 ns   20652524 ns
33   7.21102GB/s   57.6882k items/s
BM_Dataset_as_Histogram/16384        21969882 ns   21970554 ns
28   6.77843GB/s   27.1137k items/s
BM_Dataset_as_Histogram/32768        21765772 ns   21766605 ns
32   6.84194GB/s   13.6839k items/s
BM_Dataset_as_Histogram_with_slice  128666664 ns  128672616 ns
5   1.15807GB/s   75.8951k items/s

SimonHeybrock committed:
Optimize binary operation to not create VariableView if dims match.

--------------------------------------------------------------------------
Benchmark                                   Time           CPU
Iterations
--------------------------------------------------------------------------
BM_Dataset_as_Histogram/1           332504283 ns  332515409 ns
3    45.889MB/s   2.86806M items/s
BM_Dataset_as_Histogram/2           223656601 ns  223662741 ns
3   136.445MB/s   4.26389M items/s
BM_Dataset_as_Histogram/4           241634275 ns  241640825 ns
3   252.586MB/s   3.94666M items/s
BM_Dataset_as_Histogram/8           238845027 ns  238851640 ns
3   511.072MB/s   3.99275M items/s
BM_Dataset_as_Histogram/16           77832044 ns   77835651 ns
9   1.91444GB/s   7.65776M items/s
BM_Dataset_as_Histogram/32           48121594 ns   48123297 ns
15   3.09645GB/s   6.19291M items/s
BM_Dataset_as_Histogram/64           42228916 ns   42230906 ns
17    3.5285GB/s    3.5285M items/s
BM_Dataset_as_Histogram/128          31627259 ns   31628107 ns
23   4.71137GB/s   2.35568M items/s
BM_Dataset_as_Histogram/256          26752882 ns   26753186 ns
26   5.56979GB/s   1.39245M items/s
BM_Dataset_as_Histogram/512          20711493 ns   20712102 ns
34   7.19433GB/s   920.874k items/s
BM_Dataset_as_Histogram/1024         17996131 ns   17997084 ns
40   8.27923GB/s   529.871k items/s
BM_Dataset_as_Histogram/2048         17297006 ns   17297550 ns
40   8.61317GB/s   275.622k items/s
BM_Dataset_as_Histogram/4096         16482196 ns   16482495 ns
42   9.03909GB/s   144.625k items/s
BM_Dataset_as_Histogram/8192         18928742 ns   18929092 ns
38   7.86756GB/s   62.9405k items/s
BM_Dataset_as_Histogram/16384        20794920 ns   20795426 ns
34   7.16147GB/s   28.6459k items/s
BM_Dataset_as_Histogram/32768        20783723 ns   20784382 ns
34   7.16527GB/s   14.3305k items/s
BM_Dataset_as_Histogram_with_slice  117465651 ns  117469777 ns
6   1.26851GB/s   83.1331k items/s

SimonHeybrock committed:
Move Dimensions into VariableConcept for slightly better performance.

--------------------------------------------------------------------------
Benchmark                                   Time           CPU
Iterations
--------------------------------------------------------------------------
BM_Dataset_as_Histogram/1           227551239 ns  227555481 ns
3   67.0552MB/s   4.19095M items/s
BM_Dataset_as_Histogram/2           227581278 ns  227585296 ns
3   134.093MB/s    4.1904M items/s
BM_Dataset_as_Histogram/4           230352756 ns  230360912 ns
3   264.954MB/s   4.13991M items/s
BM_Dataset_as_Histogram/8           221955995 ns  221958298 ns
3    549.97MB/s   4.29664M items/s
BM_Dataset_as_Histogram/16           76084878 ns   76087537 ns
9   1.95842GB/s   7.83369M items/s
BM_Dataset_as_Histogram/32           46792823 ns   46794207 ns
15    3.1844GB/s   6.36881M items/s
BM_Dataset_as_Histogram/64           40017389 ns   40019358 ns
18   3.72349GB/s   3.72349M items/s
BM_Dataset_as_Histogram/128          31216252 ns   31216925 ns
22   4.77342GB/s   2.38671M items/s
BM_Dataset_as_Histogram/256          31053144 ns   31054021 ns
23    4.7984GB/s    1.1996M items/s
BM_Dataset_as_Histogram/512          21502696 ns   21503272 ns
33   6.92963GB/s   886.993k items/s
BM_Dataset_as_Histogram/1024         19082865 ns   19083743 ns
36    7.8078GB/s   499.699k items/s
BM_Dataset_as_Histogram/2048         18272800 ns   18273670 ns
38   8.15309GB/s   260.899k items/s
BM_Dataset_as_Histogram/4096         16891034 ns   16891362 ns
40    8.8203GB/s   141.125k items/s
BM_Dataset_as_Histogram/8192         18065254 ns   18065774 ns
36   8.24353GB/s   65.9483k items/s
BM_Dataset_as_Histogram/16384        20815084 ns   20815287 ns
33   7.15464GB/s   28.6185k items/s
BM_Dataset_as_Histogram/32768        21997570 ns   21998118 ns
33   6.76993GB/s   13.5399k items/s
BM_Dataset_as_Histogram_with_slice  117572663 ns  117575940 ns
6   1.26736GB/s    83.058k items/s

OwenArnold commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+  The major disadvantage is that we cannot have the same data type for standalone histograms and "histogram views".
+  - If option (ii.) is chosen we will still require a type to function as the current `HistogramData::Histogram`.
+    Keeping this type is one option, but another, potentially more attractive solution is to simply use `Dataset`.
+    Most of the required features for `Histogram` would also be supported by `Dataset`, and even in its current form `Dataset` provides features that are equivalent or surpassing what `HistogramData::Histogram` provides currently.
```
What features surpass current `HistogramData::Histogram`?

SimonHeybrock replied:
- Can in principle be multi dimensional.
- Can contain arbitrary types, so we could have an integer histogram, a histogram without uncertainties, a histogram with attached masking, ...
- `concatenate`
- Y and E can have units.
- X unit also stored in histogram.
- Can have a name.

OwenArnold commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+     - The disadvantages are that (a) the design feels a bit brittle and (b) we need a different type for const and mutable histograms, unless we are willing to resort to use `const_cast` and rely on the encapsulation given by `DatasetView`.
+  2. Nesting `DatasetView`.
+     - Accessing a histogram in the parent `DatasetView` would return a view by value, with its range restricted to data for the histogram.
+     - Semantically this would be similar to returning a `gsl::span` instead of a `std::vector &` in the case of `Dataset::get`.
```
👍 for `gsl::span`

OwenArnold replied:
So `DatasetView` has a non-owning pointer to the `DataSet`. How do we manage that? What happens when we perform an operation on a `Histogram` say, that generates a copy.

OwenArnold replied:
Ok, that makes sense. What about operations that do write?

SimonHeybrock replied:
`DatasetView` support writing, i.e., it is a mutable view. Is the name misleading?

OwenArnold replied:
One might think that it were immutable, but I don't see that as a major problem. What I'm wondering is what happens with something like `c = a * b` where `a` and `b` are both views. Is this operation allowed? Is a view returned as `c`. Maybe I need to look at the latest `DatasetView` implementation to understand?

SimonHeybrock replied:
Currently there are no operations for `DatasetView`. Think if it more like an iterator with multiple elements of various types.

If we were to implement such an operation, I think it should probably create a `Dataset`.

OwenArnold commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+
+The current implementation has a large number of features that are not supported in all combinations, e.g., while the prototype shows in a few examples that ragged dimensions can be supported, many other operations currently to not handle ragged dimensions.
+In general the implementation has been continued only to the point where it becomes clear that interface works and is thus capable of demonstrating that the designed concepts work.
+
```
I'm not sure if it's relevant to mention here or not, but I think that a task for the prototype will be to separate out the design noise, previous incarnations etc, from the actual final proposal. We definitely want to record the design history, but if it remains too mixed in with the actual proposal, it will be hard for a future and wider review audience to make full sense of the filtered design. You may have already considered that in the planning.

SimonHeybrock replied:
Yes. After this stage/review is finished I will do a general cleanup and remove old things. I have even considered moving `Dataset` to a new repository...? Would that be a good idea?

Further steps will be:
- Implement solutions for remaining open issues.
- Design description.
  - Potentially multiple options for some features that would require a vote or further invetigation?
- List of features (with guessed implementation effort).
- Rollout suggestions.

OwenArnold replied:
Looks like a good plan. Conclusion of rollout plan will inform estimates. PMB will definitely want those before resourcing work. Yes, moving to a new repository might be worthwhile.

OwenArnold commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+  It is determined by the variable with the highest dimensionality.
+  If tied, it is determined by the order in which variable tags are given, but if we implement the prototyped sorting...?
+
+- Do we need to support simultaneous iteration of two `Datasets`?
```
I would very much like to be able to `zip` and enumerate two datasets in general. Thinking about say a replacement for `Workspace2D` I think there are a number of algorithms that would benefit from a consistent way to do this. We would also want the same thing for `Histograms`

SimonHeybrock replied:
+1 for `zip`, though I am still unsure about how a natural solution could look like.

Note that iteration for `Dataset` is not supported due to the type-erasure, only for `DatasetView`.

OwenArnold commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+- How do distinguish (or avoid distinguishing) const and mutable versions of `DatasetView` and `Histogram`?
+- Can we avoid horrible template error messages in client code?
+- Compilations times are beginning to suffer, need some cleanup and move things out of header where possible.
+- How to generically refer to the "X" dimension, i.e., typically originally `Variable::Tof` but also anything derived from it?
```
Possibly related - how would you handle a transpose? Say in a `Workspace2D` does Spectrum Numbers then become X.

SimonHeybrock replied:
It is handled correctly and implicitly in all standard operations. As long as users to not rely on a fixed order of elements in a *multi-dimensional* iteration there should be no problem either. There is not issue in a 1D iteration, for example, when accessing a `Histogram` in `Dimension::X`, `DatasetView` will take care of on-the-fly transposition.

There is one complication however: Unless requested explicitly for a nested iteration over a histogram, `DatasetView` iterates over all dimensions. Unless client code checks whether it is multi-dimensional or not, it cannot make any assumptions.

SimonHeybrock committed:
Small cleanup.

OwenArnold reviewed (commented):
This is not a full review, but i've run out of time for the moment. Am pushing what I have reviewed already. Will review the examples asap.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+  Coordinate variables include `Coord::Tof`, `Coord::Q`, `Coord::SpectrumNumber`, `Coord::DetectorId`, `Coord::SpectrumPosition`, `Coord::DetectorPosition`, `Coord::DetectorGrouping`, `Coord::SpectrumLabel`, `Coord::RowLabel`, and `Coord::ColumnLabel`.
+
+- Add basic support for units.
+  - Variables are automatically assigned a default unit based in their tag, e.g., `Coord::X` wold have the unit "length".
```
I think the `Scale` would not be separate from the `Unit`. Need to dig more into https://www.boost.org/doc/libs/1_68_0/doc/html/boost_units.html where the introduction explains dimensions, base units, and scaled units.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+  Such variables do not exist in the `Dataset` but can be derived from other variables.
+  In the case of `Coord::SpectrumPosition` we use `Coord::DetectorPosition` and `Coord::DetectorGrouping`.
+
+- Add `operator+=` and `operator*=` for `Dataset` as examples of arithmetic operations.
```
I certain what you are referring to, but currently there are no expression templates or similar. See, e.g., the computation of the variance in `operator*=`, which is right now inefficiently creating temporaries and streaming through the data multiple times.

I would not claim that I have made a decision not to change this, but at this point I am not sure how to solve the involved complexities.

OwenArnold replied:
I was thinking along the lines of expression templates. I think this may be too complex.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+  The current implementation copies data when creating the slice --- the slice is simply a new `Dataset`.
+  It would be possible to implement slicing based on a view, avoiding the copy, however only for read-only views.
+  A mutable slice would be in conflict with the copy-on-write mechanism, since creating a mutable slice would require triggering the copy-on-write for all variables in the dataset.
+  The inefficiency of creating a new `Dataset` for every slice could be alleviated by a memory pool for the allocation of data in `Variable`.
```
I am not really familiar with `boost::pool`. From the documentation I get the impression that it is meant for allocating many small objects.

What I had in mind here was something to deal with comparitively large objects that are frequently allocated and deallocated, such as a temporary in a binary operation between vector-like objects. A previous colleague and I have used this successfully for optimizing computation with binary operations in a loop involving very large objects of fixed size. It was very simple and simply kept a pool of objects on deallocation, any new allocation (of the same size) would simply get an existing object.

In our case we cannot guarantee object sizes, but if you think of iterating over slices we would have a pattern with many identical sizes for the length of at least that iteration.

SimonHeybrock commented on doc/type-erased-prototype.md:
```diff
@@ -400,3 +410,156 @@ if (d.contains<BinMask>())
+
+- Add `DatasetIndex` to determine an index that can be used, e.g., for slicing.
+
+- Add `concatenate` for `Dataset`, a generic multi-dimensional way of merging datasets along a dimension.
```
`Dataset` is agnostic of what it contains or is meant for, so it should remain at the algorithm level.

The current implementation is not dealing with bin edges correctly, I still need to implement that (fail to concatenate unless last bin edges of first operand matches first bin edge of last operand).

SimonHeybrock committed:
Add macro for getter mixins.

SimonHeybrock committed:
Remove outdated mixins from DatasetView.

SimonHeybrock committed:
Typos.

SimonHeybrock committed:
Add benchmark exploring benefits of cache blocking.

SimonHeybrock committed:
Avoid some temporaries in Dataset::operator*=.

Gives 20-30% performance improvement.

SimonHeybrock committed:
Add Dataset::setSlice.

SimonHeybrock committed:
Fix bug in optimization of operator*=.

Multiplying with itself was broken by recent optimization.

SimonHeybrock committed:
Avoid use of VariableView when possible in slicing operations.

SimonHeybrock committed:
Reduce creation of temporaries in Variable binary ops.

SimonHeybrock committed:
Benchmarks for Dataset += and *=.

Demonstrates that current *= implementation is really slow due to bad
access patterns for propagating uncertainties.

SimonHeybrock committed:
Optimize Dataset::operator*= for matching dimensions.

Almost 10x faster.

SimonHeybrock committed:
Avoid copy in Dataset benchmark.

Intermediate conclusion:
Blocking currently 2x slower, almost half of time is spent in `slice`
and `setSlice` and for some reason the `doWork` does not speed up when
data is in cache.

SimonHeybrock committed:
Avoid copying in Variable::setDimensions() if possible.

SimonHeybrock committed:
Fail on empty match in Dataset::extract.

SimonHeybrock committed:
Avoid some issues in Dataset benchmark.

SimonHeybrock committed:
Attempt to auto-vectorize operator*=.

SimonHeybrock committed:
Align data in Variable.

Benchmark for multiplication of Dataset shows performance 10x above
memory bandwidth when data is in cache (with multiple threads),
indicating that the overhead from Dataset (type erasure, dict-like
structure, ...) is not dominant?

-------------------------------------------------------------------------------------
Benchmark                                              Time
CPU Iterations
-------------------------------------------------------------------------------------
BM_Dataset_multiply/2                               2148 ns       2148
ns     262786   42.6312GB/s   909.466k items/s
BM_Dataset_multiply/4                               3560 ns       3560
ns     197564   51.4301GB/s   1097.17k items/s
BM_Dataset_multiply/8                               7884 ns       7885
ns      89777   46.4455GB/s   990.838k items/s
BM_Dataset_multiply/16                             18767 ns      18768
ns      37122    39.026GB/s   832.555k items/s
BM_Dataset_multiply/32                             36525 ns      36527
ns      19195   40.1033GB/s   855.538k items/s
BM_Dataset_multiply/64                             71789 ns      71792
ns       9699   40.8083GB/s   870.576k items/s
BM_Dataset_multiply/128                           143051 ns     143057
ns       4865   40.9584GB/s   873.779k items/s
BM_Dataset_multiply/256                           290120 ns     290132
ns       2409   40.3911GB/s   861.677k items/s
BM_Dataset_multiply/512                          1030315 ns    1030348
ns        678   22.7472GB/s   485.273k items/s
BM_Dataset_multiply/1024                         2733307 ns    2733364
ns        259   17.1492GB/s    365.85k items/s
BM_Dataset_multiply/2048                         5506082 ns    5506346
ns        129   17.0258GB/s   363.217k items/s
BM_Dataset_multiply/4096                        11473516 ns   11473936
ns         62   16.3414GB/s   348.616k items/s
BM_Dataset_multiply/8192                        22414931 ns   22415746
ns         31   16.7293GB/s   356.892k items/s
BM_Dataset_multiply/2/real_time/threads:4            822 ns       3287
ns     807904   111.424GB/s   2.32134M items/s
BM_Dataset_multiply/4/real_time/threads:4           1220 ns       4882
ns     417068   150.031GB/s   3.12565M items/s
BM_Dataset_multiply/8/real_time/threads:4           2692 ns      10769
ns     211512   136.029GB/s   2.83393M items/s
BM_Dataset_multiply/16/real_time/threads:4          6698 ns      26792
ns      95580   109.354GB/s   2.27821M items/s
BM_Dataset_multiply/32/real_time/threads:4         13073 ns      52291
ns      43472   112.055GB/s   2.33448M items/s
BM_Dataset_multiply/64/real_time/threads:4         26864 ns     107459
ns      24552   109.056GB/s   2.27201M items/s
BM_Dataset_multiply/128/real_time/threads:4        54260 ns     217045
ns      12552   107.986GB/s   2.24971M items/s
BM_Dataset_multiply/256/real_time/threads:4       232754 ns     930964
ns       2380   50.3482GB/s   1074.09k items/s
BM_Dataset_multiply/512/real_time/threads:4      1078132 ns    4312596
ns        832    21.739GB/s   463.765k items/s
BM_Dataset_multiply/1024/real_time/threads:4     1444659 ns    5778656
ns        512   32.4471GB/s   692.205k items/s
BM_Dataset_multiply/2048/real_time/threads:4     2611347 ns   10445617
ns        320    35.901GB/s   765.888k items/s
BM_Dataset_multiply/4096/real_time/threads:4     7845748 ns   31382957
ns         84   23.8983GB/s    509.83k items/s
BM_Dataset_multiply/8192/real_time/threads:4    17138461 ns   68553668
ns         48   21.8806GB/s   466.786k items/s
BM_Dataset_multiply/2/real_time/threads:8            393 ns       3140
ns    1619632   233.254GB/s   4.85946M items/s
BM_Dataset_multiply/4/real_time/threads:8            795 ns       6361
ns     629792   230.283GB/s   4.79755M items/s
BM_Dataset_multiply/8/real_time/threads:8           1310 ns      10481
ns     402048   279.521GB/s   5.82335M items/s
BM_Dataset_multiply/16/real_time/threads:8          3104 ns      24830
ns     195008    235.99GB/s   4.91647M items/s
BM_Dataset_multiply/32/real_time/threads:8          6189 ns      49517
ns      81264   236.668GB/s   4.93058M items/s
BM_Dataset_multiply/64/real_time/threads:8         13301 ns     106407
ns      53664   220.266GB/s   4.58887M items/s
BM_Dataset_multiply/128/real_time/threads:8        56192 ns     449432
ns      12448   104.273GB/s   2.17236M items/s
BM_Dataset_multiply/256/real_time/threads:8       321959 ns    2575640
ns       2288   36.3982GB/s   776.495k items/s
BM_Dataset_multiply/512/real_time/threads:8       672395 ns    5378982
ns        848   34.8567GB/s    743.61k items/s
BM_Dataset_multiply/1024/real_time/threads:8     1789437 ns   14312010
ns        408   26.1954GB/s   558.835k items/s
BM_Dataset_multiply/2048/real_time/threads:8     2094371 ns   16755450
ns        240   44.7628GB/s   954.941k items/s
BM_Dataset_multiply/4096/real_time/threads:8     6370479 ns   50962740
ns        104   29.4326GB/s   627.896k items/s
BM_Dataset_multiply/8192/real_time/threads:8    14262582 ns  114102571
ns         48   26.2926GB/s   560.908k items/s
BM_Dataset_multiply/2/real_time/threads:12           262 ns       3144
ns    2638764   349.443GB/s   7.28006M items/s
BM_Dataset_multiply/4/real_time/threads:12           480 ns       5758
ns    1137912   381.595GB/s    7.9499M items/s
BM_Dataset_multiply/8/real_time/threads:12           852 ns      10212
ns     845316   429.917GB/s   8.95661M items/s
BM_Dataset_multiply/16/real_time/threads:12         2574 ns      30885
ns     216132   284.525GB/s    5.9276M items/s
BM_Dataset_multiply/32/real_time/threads:12         5572 ns      66665
ns     106440   262.893GB/s   5.47693M items/s
BM_Dataset_multiply/64/real_time/threads:12         9888 ns     118538
ns      68640   296.294GB/s    6.1728M items/s
BM_Dataset_multiply/128/real_time/threads:12       62117 ns     745342
ns      12504   94.3285GB/s   1.96518M items/s
BM_Dataset_multiply/256/real_time/threads:12      371976 ns    4461880
ns       2484   31.5041GB/s   672.087k items/s
BM_Dataset_multiply/512/real_time/threads:12      675927 ns    8082099
ns        780   34.6746GB/s   739.725k items/s
BM_Dataset_multiply/1024/real_time/threads:12    1333619 ns   15999629
ns        444   35.1487GB/s   749.839k items/s
BM_Dataset_multiply/2048/real_time/threads:12    2717016 ns   32593554
ns        192   34.5048GB/s   736.102k items/s
BM_Dataset_multiply/4096/real_time/threads:12    6157815 ns   73857734
ns         96   30.4491GB/s   649.581k items/s
BM_Dataset_multiply/8192/real_time/threads:12   14254706 ns  170879966
ns         72   26.3071GB/s   561.218k items/s
BM_Dataset_multiply/2/real_time/threads:24           204 ns       4850
ns    2926512   449.639GB/s   9.36748M items/s
BM_Dataset_multiply/4/real_time/threads:24           401 ns       9465
ns    1702320   456.878GB/s   9.51829M items/s
BM_Dataset_multiply/8/real_time/threads:24           923 ns      22053
ns     775680    396.61GB/s   8.26271M items/s
BM_Dataset_multiply/16/real_time/threads:24         1842 ns      43994
ns     394056   397.596GB/s   8.28326M items/s
BM_Dataset_multiply/32/real_time/threads:24         3630 ns      86791
ns     176592   403.521GB/s   8.40668M items/s
BM_Dataset_multiply/64/real_time/threads:24        13121 ns     310228
ns      42288    223.29GB/s   4.65188M items/s
BM_Dataset_multiply/128/real_time/threads:24      182569 ns    4349361
ns       4008   32.0941GB/s   684.674k items/s
BM_Dataset_multiply/256/real_time/threads:24      419719 ns   10033492
ns       1848   27.9205GB/s   595.636k items/s
BM_Dataset_multiply/512/real_time/threads:24      620693 ns   14739685
ns       1296   37.7602GB/s   805.551k items/s
BM_Dataset_multiply/1024/real_time/threads:24     979297 ns   23278434
ns        720    47.866GB/s   1021.14k items/s
BM_Dataset_multiply/2048/real_time/threads:24    1938041 ns   46216854
ns        360   48.3736GB/s   1031.97k items/s
BM_Dataset_multiply/4096/real_time/threads:24    3883027 ns   92717103
ns        192   48.2871GB/s   1030.12k items/s
BM_Dataset_multiply/8192/real_time/threads:24    7652822 ns  182517514
ns         96   49.0015GB/s   1045.37k items/s

SimonHeybrock committed:
Updated documentation.

SimonHeybrock committed:
Typos.

SimonHeybrock committed:
Fix bad alignment specification.

SimonHeybrock committed:
Clarify some details in notes.

SimonHeybrock committed:
Add GPL licence.

SimonHeybrock commented:
I have moved the `Dataset` prototype to a new repository, where I merged this branch and deleted the older prototypes as well as the old `Histogram` implementation and "ragged" dimension support: https://github.com/mantidproject/dataset. Development will continue there from now on.

For now, I will leave this PR open in case there are further comments.
